src/Day17/Solution.hs

```diff
@@ -0,0 +1,133 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Day17.Solution
+  ( CubeState (..),
+    Point3D (..),
+    Point4D (..),
+    executeCycles,
+    parsePocketDimension3D,
+    parsePocketDimension4D,
+    part1,
+    part2,
+  )
+where
+
+import Advent.Utils
+  ( fromRightOrShowError,
+    isBetween,
+    occurrences,
+  )
+import Data.Function ((&))
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import Text.Parsec
+
+part1 :: String -> String
+part1 = show . occurrences Active . executeCycles 6 . fromRightOrShowError . parsePocketDimension3D
+
+part2 :: String -> String
+part2 = show . occurrences Active . executeCycles 6 . fromRightOrShowError . parsePocketDimension4D
+
+class Pocket a where
+  from2D :: (Int, Int) -> a
+  offset :: a -> a -> a
+  neighbors :: a -> Set.Set a
+
+newtype Point3D = Point3D (Int, Int, Int) deriving (Eq, Ord, Show)
+
+instance Pocket Point3D where
+  from2D (x, y) = Point3D (x, y, 0)
+  Point3D (a, b, c) `offset` Point3D (x, y, z) = Point3D (a + x, b + y, c + z)
+  neighbors point =
+    Set.fromList
+      [ point `offset` Point3D (x, y, z)
+        | (x, y, z) <- offsets3D,
+          (x, y, z) /= (0, 0, 0)
+      ]
+
+newtype Point4D = Point4D (Int, Int, Int, Int) deriving (Eq, Ord, Show)
+
+instance Pocket Point4D where
+  from2D (x, y) = Point4D (x, y, 0, 0)
+  Point4D (a, b, c, d) `offset` Point4D (x, y, z, w) = Point4D (a + x, b + y, c + z, d + w)
+  neighbors point =
+    Set.fromList
+      [ point `offset` Point4D (x, y, z, w)
+        | (x, y, z, w) <- offsets4D,
+          (x, y, z, w) /= (0, 0, 0, 0)
+      ]
```

Comment:
This is my first attempt at writing a class in Haskell. 🎉

I think it makes sense here for a couple of reasons. My first instinct is usually dependency injection -- pass in functions like neighbors as dependencies. However, I couldn't figure out how I would "type" these dependencies to work on both 3-tuples and 4-tuples, which are very different things in Haskell.

I think the classes make sense.